src/classes/RD2_Schedule_SVC.cls

```diff
@@ -0,0 +1,185 @@
+        rdsched.InstallmentFrequency__c = rd.InstallmentFrequency__c;
+        rdsched.InstallmentPeriod__c = rd.npe03__Installment_Period__c;
+        rdsched.PaymentMethod__c = rd.PaymentMethod__c;
+        rdsched.RecurringDonation__c = rd.Id;
+
+        if (rdsched.InstallmentPeriod__c == MONTHLY) {
+            rdsched.DayOfMonth__c = getDayOfMonth(rd);
+        }
+        schedulesToInsert.add(rdsched);
+
+        if (isFirstAndFifteenth(rd)) {
+            rdsched = rdsched.clone();
+            rdsched.DayOfMonth__c = FIFTEENTH;
+            schedulesToInsert.add(rdsched);
+        }
+
+        for (RecurringDonationSchedule__c rds : schedulesToInsert) {
+            rds.StartDate__c = calculateStartDate(rds, rd);
+        }
+
+        return this;
+    }
+
+    // TODO: Update schedule to reflect recurringDonation - W-036113 (Placeholder)
+    public RD2_Schedule_SVC updateSchedule(npe03__Recurring_Donation__c rd, List<RecurringDonationSchedule__c> rdScheds) {
+        return this;
+    }
+
+    // TODO: Will return next Donation date - W-036251 (Placeholder)
+    public Date getNextDonationDate(RecurringDonationSchedule__c rdSched) {
+        return null;
+    }
+
+    // TODO: Will return an inner class (Placeholder - RDScheduleDates)
+    public List<RDScheduleDates> visualizeSchedule(npe03__Recurring_Donation__c rd, List<RecurringDonationSchedule__c> rdScheds) {
+        return new List<RDScheduleDates>();
+    }
+
+    // TODO: Placeholder internal class used for visualization)
+    public class RDScheduleDates {
+        public npe03__Recurring_Donation__c recurringDonation;
+        public List<RecurringDonationSchedule__c> rdSchedules;
+        public Date donationDate;
+    }
+
+    /***
+    * @description Calculates start date for a recurring donation schedule
+    * @param rdSchedule Recurring Donation schedule
+    * @param rd Recurring Donation schedule
+    * @return Date
+    */
+    private Date calculateStartDate(RecurringDonationSchedule__c rdSchedule, npe03__Recurring_Donation__c rd) {
+        Date retDate;
+
+        Date minSD = rd.StartDate__c > rd.npe03__Date_Established__c ?
+                     rd.StartDate__c :
+                     rd.npe03__Date_Established__c;
+
+        retDate = minSD > Date.today() ?
+                  minSD :
+                  Date.today();
+
+        if (rdSchedule.InstallmentPeriod__c == MONTHLY) {
+            Integer lastDay = Date.daysInMonth(retDate.year(), retDate.month());
+            retDate = Date.newInstance(retDate.year(), retDate.month(), rdSchedule.DayOfMonth__c == LAST_DAY_OF_MONTH ?
+                                                                        lastDay :
+                                                                        Integer.valueOf(rdSchedule.DayOfMonth__c));
+            retDate = retDate >= Date.today() ?
+                      retDate :
+                      retDate.addMonths(1);
```

Comment:
I believe this assumes a Monthly schedule. It might be better to add logic to handle other schedule types such as Daily, Weekly and Yearly this way it adds the appropriate interval.

Reply:
This section of code should check for either monthly or 1st and 15th in period, so I need to fix that instead of just testing for monthly, but other schedule types should already be handled earlier in the method.  Do you think there is adjustment needed in the logic?

Reply:
Good point. I wonder if we need to eventually allow for a grace period on these. For example, what if the record was inserted into the BDI object on 10/21, but not processed - and thus not inserted into the RD object - until 10/22. In that case, the intent is for the record to have a start date of 10/21 even though it's not actually being processed until 10/22. More than likely, BDI would also be inserting/matching to the first installment.

Thinking about these use cases more, I would probably recommend that we don't push it out to the next month at all if the StartDate is before the current date. I realize this doesn't match the work item A/C's though. Let's talk through it during stand-up.